core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -1700,19 +1700,34 @@ class SparkContext(config: SparkConf) extends Logging {
    * Adds a JAR dependency for all tasks to be executed on this SparkContext in the future.
    * The `path` passed can be either a local file, a file in HDFS (or other Hadoop-supported
    * filesystems), an HTTP, HTTPS or FTP URI, or local:/path for a file on every worker node.
+   * If addToCurrentClassLoader is true, attempt to add the new class to the current threads' class
+   * loader. In general adding to the current threads' class loader will impact all other
+   * application threads unless they have explicitly changed their class loader.
    */
   def addJar(path: String) {
+    addJar(path, false)
+  }
+
+  def addJar(path: String, addToCurrentClassLoader: Boolean) {
     if (path == null) {
       logWarning("null specified as parameter to addJar")
     } else {
       var key = ""
-      if (path.contains("\\")) {
+
+      val uri = if (path.contains("\\")) {
         // For local paths with backslashes on Windows, URI throws an exception
-        key = env.rpcEnv.fileServer.addJar(new File(path))
```

Comment:
So this change gets the URI for the windows URI which is used later on to construct a File instance.  That should allow the windows special case to work.